hazelcast/src/main/java/com/hazelcast/cache/impl/operation/CacheReplicationOperation.java

```diff
@@ -102,8 +104,16 @@ public void beforeRun() throws Exception {
     public void run() throws Exception {
         ICacheService service = getService();
         for (Map.Entry<String, Map<Data, CacheRecord>> entry : data.entrySet()) {
-            ICacheRecordStore cache = service.getOrCreateRecordStore(entry.getKey(), getPartitionId());
-            cache.reset();
+            // establish thread-local context for this cache's tenant application before possibly creating records
+            // This is so CDI / JPA / EJB methods can be called from other than JavaEE threads
+            Closeable tenantContext = getTenantControl(service.getCacheConfig(entry.getKey())).setTenant(true);
+            ICacheRecordStore cache;
+            try {
+                cache = service.getOrCreateRecordStore(entry.getKey(), getPartitionId());
+                cache.reset();
+            } finally {
+                tenantContext.close();
```

Comment:
Out of curiosity, why don't we close the context after putting records?

Reply:
Hm, well spotted! there are invocations in `cache.putRecord` which may result in serialization (near cache invalidations, event journal updates etc), so population of cache records must happen within tenant context.